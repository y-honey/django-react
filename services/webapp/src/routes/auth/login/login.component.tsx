import React from 'react';

import { FormattedMessage } from 'react-intl';
import { LoginForm } from '../../../shared/components/auth/loginForm';
import { H1 } from '../../../theme/typography';
import { Container } from './login.styles';

export const Login = () => {
  return (
    <Container>
      <H1>
        <FormattedMessage defaultMessage="Login" description="Auth / login / heading" />
      </H1>
      <LoginForm />
    </Container>
  );
};
