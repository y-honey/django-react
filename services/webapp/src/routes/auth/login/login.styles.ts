import styled from 'styled-components';

export const Container = styled.div`
  text-align: center;
`;

export const Links = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;
