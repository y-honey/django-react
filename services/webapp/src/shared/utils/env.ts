export const getEnvNumber = (key: keyof NodeJS.ProcessEnv) => {
  const value = process.env[key];
  if (value === undefined) return;
  const number = Number(value);
  if (isNaN(number)) return;
  return number;
};
