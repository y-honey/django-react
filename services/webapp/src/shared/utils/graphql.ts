import { UnknownObject } from './types';

export type ConnectionType<ITEM> = {
  readonly edges: ReadonlyArray<{
    readonly node: ITEM | null;
  } | null>;
} | null;

export const mapConnection = <ITEM, RETURN>(
  callback: (item: ITEM) => RETURN,
  query: ConnectionType<ITEM>
): RETURN[] => {
  const existingNodes: RETURN[] = [];
  const edges = query?.edges ?? [];

  for (const edge of edges) {
    if (edge?.node) {
      existingNodes.push(callback(edge.node));
    }
  }

  return existingNodes;
};

export type ExtractNodeType<
  T extends {
    readonly edges: ReadonlyArray<{
      readonly node: ITEM | null;
    } | null>;
  } | null,
  ITEM extends UnknownObject = UnknownObject
> = NonNullable<NonNullable<NonNullable<NonNullable<T>['edges']>[number]>['node']>;
