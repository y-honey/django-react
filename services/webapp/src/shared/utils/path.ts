import { map } from 'ramda';

export const getLocalePath = (p: string) => `/:lang/${p}`;

export const nestedPath = <T extends string>(root: string, nestedRoutes: Record<T, string>) => {
  const absoluteNestedUrls = map<Record<T, string>, Record<T, string>>((value) => root + '/' + value, nestedRoutes);
  const paths = {
    index: `${root}/*`,
    ...absoluteNestedUrls
  };
  return {
    ...paths,
    getRelativeUrl: (route: T) => nestedRoutes[route],
    getLocalePath: (route: T) => getLocalePath(paths[route]),
  };
};
