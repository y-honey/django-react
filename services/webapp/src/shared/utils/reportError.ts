import * as Sentry from '@sentry/react';

/**
 * Generic function for error reporting.
 */
export const reportError = (error: any): void => {
  if (process.env.NODE_ENV !== 'test') {
    console.error(error); //eslint-disable-line
  }
  if (!error?.response) {
    Sentry.captureException(error);
  }
};
