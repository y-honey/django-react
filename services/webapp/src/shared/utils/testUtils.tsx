import { ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { identity } from 'ramda';
import { Provider } from 'react-redux';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Route, Routes } from 'react-router';
import { unstable_HistoryRouter as HistoryRouter } from 'react-router-dom';
import { IntlProvider } from 'react-intl';
import { produce } from 'immer';
import { RelayEnvironmentProvider } from 'react-relay';
import { createMockEnvironment, RelayMockEnvironment } from 'relay-test-utils';

import { DEFAULT_LOCALE, translationMessages, TranslationMessages } from '../../app/config/i18n';
import { store as fixturesStore } from '../../mocks/store';
import createReducer, { GlobalState } from '../../app/config/reducers';
import { ResponsiveThemeProvider } from '../../app/providers/responsiveThemeProvider';
import { CommonQuery } from '../../app/providers/commonQuery';
import { fillCommonQueryWithUser } from './commonQuery';

export const PLACEHOLDER_TEST_ID = 'content';
export const PLACEHOLDER_CONTENT = <span data-testid="content">content</span>;

export const spiedHistory = (route = `/${DEFAULT_LOCALE}`) => {
  const history = createMemoryHistory({ initialEntries: [route] });
  history.push = jest.fn();
  return {
    history,
    pushSpy: history.push as jest.Mock,
  };
};

export interface ContextData {
  router?: {
    url?: string;
    routePath?: string;
    history?: MemoryHistory;
    children?: ReactNode;
  };
  store?: GlobalState | ((draft: GlobalState) => void);
  messages?: TranslationMessages;
  relayEnvironment?: RelayMockEnvironment | ((env: RelayMockEnvironment) => void);
}

export interface ProvidersWrapperProps {
  children: ReactNode;
  context?: ContextData;
}

export const ProvidersWrapper = ({ children, context = {} }: ProvidersWrapperProps) => {
  const { router = {}, messages } = context;
  const { url = `/${DEFAULT_LOCALE}`, routePath = '/:lang/', history } = router;

  const routerHistory: MemoryHistory = history ?? createMemoryHistory({ initialEntries: [url] });

  const intlProviderMockProps = {
    locale: DEFAULT_LOCALE,
    messages: messages ?? translationMessages[DEFAULT_LOCALE],
  };

  const store = (() => {
    if (!context?.store) return fixturesStore;
    if (typeof context.store === 'function') return produce(fixturesStore, context.store);
    return context.store;
  })();

  const relayEnvironment = (() => {
    if (context?.relayEnvironment === undefined) {
      const env = createMockEnvironment();
      fillCommonQueryWithUser(env);
      return env;
    }
    if (typeof context.relayEnvironment === 'function') {
      const mockEnvironment = createMockEnvironment();
      context.relayEnvironment(mockEnvironment);
      return mockEnvironment;
    }
    return context.relayEnvironment;
  })();

  return (
    <HistoryRouter history={routerHistory}>
      <HelmetProvider>
        <ResponsiveThemeProvider>
          <RelayEnvironmentProvider environment={relayEnvironment}>
            <CommonQuery>
              <IntlProvider {...intlProviderMockProps}>
                <Provider store={createStore(createReducer(), produce(store, identity))}>
                  <Routes>
                    <Route path={routePath} element={children}>
                      {router.children}
                    </Route>
                  </Routes>
                </Provider>
              </IntlProvider>
            </CommonQuery>
          </RelayEnvironmentProvider>
        </ResponsiveThemeProvider>
      </HelmetProvider>
    </HistoryRouter>
  );
};

export function makeContextRenderer<T>(
  component: (props: T | Record<string, never>) => ReactElement,
  baseContext: ContextData = {}
) {
  return (props?: T, context?: ContextData) => {
    return render(component(props ?? {}), {
      wrapper: ({ children }) => (
        <ProvidersWrapper context={{ ...baseContext, ...context }}>{children}</ProvidersWrapper>
      ),
    });
  };
}

export function makePropsRenderer<T>(component: (props: T | Record<string, never>) => ReactElement) {
  return (props?: T) => render(component(props ?? {}));
}

// using `screen.getByText(matchTextContent('hello world'))` will match <div>hello <span>world</span></div>
export const matchTextContent = (text: string | RegExp) => (_: unknown, node: Element | null | undefined) => {
  const hasText = (node: Element) => {
    const isPlainString = typeof text === 'string';
    const nodeContent = node.textContent ?? '';
    return isPlainString ? nodeContent === text : new RegExp(text).test(nodeContent);
  };
  const childrenDontHaveText = Array.from(node?.children ?? []).every((child) => !hasText(child));
  return Boolean(node && hasText(node) && childrenDontHaveText);
};

export const connectionFromArray = <T extends Record<string, any>>(arr: T[] = []) => {
  if (!arr || arr.length === 0) {
    return {
      edges: [],
      totalCount: 0,
      count: 0,
      endCursorOffset: 0,
      startCursorOffset: 0,
      pageInfo: {
        hasNextPage: false,
        hasPreviousPage: false,
      },
    };
  }

  return {
    edges: arr.map((node) => ({ node })),
    totalCount: arr.length,
    count: arr.length,
    endCursorOffset: arr.length,
    startCursorOffset: 0,
    pageInfo: {
      hasNextPage: false,
      hasPreviousPage: false,
    },
  };
};

export const packHistoryArgs = (pathname: string, other = {}) => [
  expect.objectContaining({ pathname, ...other }),
  undefined,
];
