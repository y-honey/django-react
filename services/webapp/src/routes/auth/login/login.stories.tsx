import { Story } from '@storybook/react';
import { withProviders } from '../../../shared/utils/storybook';
import { Login } from './login.component';

const Template: Story = () => {
  return <Login />;
};

export default {
  title: 'Routes/Auth/Login',
  component: Login,
};

export const Default = Template.bind({});
Default.decorators = [withProviders({})];
