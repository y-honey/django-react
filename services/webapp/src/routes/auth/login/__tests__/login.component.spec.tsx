import React from 'react';

import { makeContextRenderer } from '../../../../shared/utils/testUtils';
import { Login } from '../login.component';

describe('Login: Component', () => {
  const component = () => <Login />;
  const render = makeContextRenderer(component);

  it('should render correctly', () => {
    const { container } = render();
    expect(container.firstChild).toMatchSnapshot();
  });
});
