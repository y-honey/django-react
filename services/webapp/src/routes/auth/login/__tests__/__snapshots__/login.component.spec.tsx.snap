// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login: Component should render correctly 1`] = `
.c2 {
  padding: 8px 16px;
  border: 1px solid #80809B;
  color: #42f272;
}

.c1 {
  font-family: OpenSans;
  font-weight: bold;
  color: #000000;
}

.c0 {
  text-align: center;
}

<div
  class="c0"
>
  <h1
    class="c1"
  >
    Login
  </h1>
  <form
    class=""
  >
    <div
      class=""
    >
      <input
        class=""
        name="email"
        placeholder="Email"
        type="email"
      />
      <p
        class=""
      />
    </div>
    <div
      class=""
    >
      <input
        class=""
        name="password"
        placeholder="Password"
        type="password"
      />
      <p
        class=""
      />
    </div>
    <button
      class="c2"
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
